presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCache.java

```diff
@@ -0,0 +1,90 @@
+import org.testng.annotations.Test;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.airlift.testing.Assertions.assertGreaterThan;
+import static io.prestosql.tempto.assertions.QueryAssert.Row.row;
+import static io.prestosql.tempto.assertions.QueryAssert.assertThat;
+import static io.prestosql.tempto.query.QueryExecutor.query;
+import static io.prestosql.tests.TestGroups.HIVE_CACHE;
+import static java.lang.Thread.sleep;
+import static org.testng.Assert.assertEquals;
+
+public class TestCache
+        extends ProductTest
+{
+    @Test(groups = HIVE_CACHE)
+    public void testReadFromCache()
+            throws InterruptedException
+    {
+        String nonCachedTableName = "hive.default.test_cache_read";
+        String cachedTableName = "hivecached.default.test_cache_read";
+
+        query("DROP TABLE IF EXISTS " + nonCachedTableName);
+        query("CREATE TABLE " + nonCachedTableName + " WITH (format='ORC') AS SELECT 'Hello world' as col");
+
+        QueryResult beforeCacheStats = getCacheStats();
+        long beforeCachedReads = getCachedReads(beforeCacheStats);
+        long beforeRemoteReads = getRemoteReads(beforeCacheStats);
+
+        assertThat(query("SELECT * FROM " + cachedTableName))
+                .containsExactly(row("Hello world"));
+
+        // stats are propagated asynchronously, wait for them
+        sleep(1000);
+        QueryResult firstCacheStats = getCacheStats();
+        long firstCachedReads = getCachedReads(firstCacheStats);
+        long firstRemoteReads = getRemoteReads(firstCacheStats);
+
+        // first query via caching catalog should fetch remote data
+        assertGreaterThan(firstRemoteReads, beforeRemoteReads);
+        assertEquals(firstCachedReads, beforeCachedReads);
+
+        assertThat(query("SELECT * FROM " + cachedTableName))
+                .containsExactly(row("Hello world"));
+
+        // stats are propagated asynchronously, wait for them
+        sleep(1000);
+        QueryResult secondCacheStats = getCacheStats();
+        long secondCachedReads = getCachedReads(secondCacheStats);
+        long secondRemoteReads = getRemoteReads(secondCacheStats);
+
+        // second query via caching catalog should read exclusively from cache
+        assertEquals(secondRemoteReads, firstRemoteReads);
+        assertGreaterThan(secondCachedReads, firstCachedReads);
+
+        query("DROP TABLE " + nonCachedTableName);
+    }
+
+    private QueryResult getCacheStats()
+    {
+        return query("SELECT cachedreads, remotereads FROM " +
+                "jmx.current.\"rubix:catalog=hivecached,name=stats\" WHERE node = 'presto-worker'");
+    }
+
+    private long getCachedReads(QueryResult queryResult)
+    {
+        return (Long) getOnlyElement(queryResult.rows()).get(0);
+    }
+
+    private long getRemoteReads(QueryResult queryResult)
+    {
+        return (Long) getOnlyElement(queryResult.rows()).get(1);
+    }
```

Comment:
`getCachedReads` method doesn't really know what it is doing, it could be called `getFirstColumn`
Similarly `getRemoteReads` could be called `getSecondColumn`.

Use `tryFindColumnIndex` to map columns by name.